### DISCUSSION TITLE
uuid as primary key

### DISCUSSION THREAD

**Discussion by josejachuf** [emmett-framework/emmett#398] [Ideas]
I thought it may be useful to add, in future versions, some parameter to set UUID as a primary key instead of using Integer. I know that it is very complicated

**Comment by gi0baro** [comment]
@josejachuf it is not.
It's already included in 2.4 (ref: https://github.com/emmett-framework/emmett/pull/394/commits/6b69f642c2aafd6a6009934d746ff53141c7a29b), where you will be able to do something like this:

```python
import uuid
from emmett.orm import Model, Field

class ExampleModel(Model):
    id = Field.string(default=lambda: uuid.uuid4().hex)

```

Hopefully I'll be able to release it before Christmas.

**Reply by josejachuf** [reply]
Thanks very much @gi0baro 

That's great! How will affect using a PK String to Emmett-REST? Will something break?

**Reply by gi0baro** [reply]
@josejachuf it would need a small update to customise the `id_path` url per-module, but apart from that it would work.

**Reply by josejachuf** [reply]
You refer to configuration level changing:
 
`app.config.REST.base_id_path = '/<int:rid>'`

by

`app.config.REST.base_id_path = '/<str:rid>' `

Or do you mean that it is necessary to modify the extension? 

Jose

**Reply by gi0baro** [reply]
@josejachuf that's the ready-to-go parameter. What I meant is that this configuration is across all modules, so I expect people to want to mix strings and integers for ids, an thus we need the same parameter also for the single modules.

**Reply by josejachuf** [reply]
Thanks. I will try

**Reply by gi0baro** [reply]
Quick update: 2.4 branch is merged in master, I'm not ready for a release yet, but I expect it in the next couple of weeks. The documentation is already on track for this: https://github.com/emmett-framework/emmett/blob/bba4354e497a3bd9d9c88e12f2cf4ad7f1d9d017/docs/orm/advanced.md#custom-primary-keys

**Comment by gi0baro** [comment]
@josejachuf this is now fully supported by Emmett 2.4 (see https://emmett.sh/docs/2.4.x/orm/advanced#custom-primary-keys) and REST extension 1.3 (https://github.com/emmett-framework/rest/releases/tag/v1.3.0)

**Reply by josejachuf** [reply]
Awesome! Thanks very much,
